### PR TITLE
Cleanup macro expansion and improve diagnostics

### LINKED ISSUE #33709: confusing error message with concat_idents!() use in type position
I tried to use `concat_idents!` to construct the name of a type in a macro. I discovered this wasn't possible, but the error message for it doesn't seem right.

```
cam@zoo /tmp $ cat m.rs
#![feature(concat_idents)]
#![feature(type_macros)]
pub enum YZ { }
macro_rules! test {
    ($x:ident, $y:ident) => (
        pub struct $x {
            pub ptr: *mut concat_idents!($y, Z),
        }
    )
}
test!(X, Y);
cam@zoo /tmp $ multirust run nightly rustc m.rs
m.rs:7:27: 7:40 error: non-expression macro in expression position: concat_idents
m.rs:7             pub ptr: *mut concat_idents!($y, Z),
                                 ^~~~~~~~~~~~~
error: aborting due to previous error
cam@zoo /tmp $ multirust run nightly rustc --version
rustc 1.10.0-nightly (cd6a40017 2016-05-16)
```

Maybe the "non-expression" and "expression" need to be swapped around?

CC @nrc

### PR DESCRIPTION
**Pull request by jseyfried**
Cleanup macro expansion and improve diagnostics. Fixes #33709.
r? @nrc

### CODE REVIEW

**Comment by nrc** [comment]
Damn, @jseyfied, you've been busy :-)

File: src/libsyntax/ext/expand.rs [hunk 1]
```diff
@@ -35,59 +35,47 @@ use std_inject;
...
 
-// this function is called to detect use of feature-gated or invalid attributes
-// on macro invoations since they will not be detected after macro expansion
-fn check_attributes(attrs: &[ast::Attribute], fld: &MacroExpander) {
-    for attr in attrs.iter() {
-        feature_gate::check_attribute(&attr, &fld.cx.parse_sess.span_diagnostic,
-                                      &fld.cx.parse_sess.codemap(),
-                                      &fld.cx.ecfg.features.unwrap());
-    }
+trait MacroGenerable: Sized {
```
Comment (nrc) [inline src/libsyntax/ext/expand.rs]: Could you comment on what this trait is for, what should implement it, etc.

**Comment by nrc** [comment]
r+ with the extra comment

**Comment by jseyfried** [comment]
@nrc I added some comments.

**Pull request merged**